Implement `of` (#331)
TuringLang/JuliaBUGS.jl: of_design_doc.md

- `of(Real, lower, upper)` → `OfReal{Val{lower}, Val{upper}}` - Bounded real numbers
- `of(Int)` → `OfInt{Nothing, Nothing}` - Unbounded integers
- `of(Int, lower, upper)` → `OfInt{Val{lower}, Val{upper}}` - Bounded integers
- `@of(field1=..., field2=...)` → `OfNamedTuple{(:field1, :field2), Tuple{Type1, Type2}}` - Named tuples (use @of macro only)
- `of(...; constant=true)` → `OfConstantWrapper{T}` - Marks a type as constant/hyperparameter (only supported for `Int` and `Real` types)

### 2. Type Parameter Encoding

The system encodes extra useful information into type parameters:
- **Dimensions**: Stored as tuple type parameters (e.g., `(3, 4)` for a 3×4 matrix)
- **Bounds**: Encoded using `Val{x}` for numeric bounds or `Nothing` for unbounded
- **Symbolic references**: Encoded using `SymbolicRef{:symbol}` for referencing other fields
- **Arithmetic expressions**: Encoded using `SymbolicExpr{expr}` for expressions like `n+1`, `2*n`, etc. Division operations must result in integers for array dimensions.
- **Field names**: Stored as tuple of symbols in `OfNamedTuple`
- **Element types**: Preserved as type parameters for arrays and nested structures

### 3. Operations on Types

> - `T(;kwargs...)` where `T<:OfType` - Create instances with specified constants (returns values, not types)

Reviewer: ```suggestion
- `T(default_value=0.;kwargs...)` where `T<:OfType` - Create instances with specified constants (returns values, not types), and initialise all element values to `default_value`, e.g. `T(default_value=0.;kwargs...)` initialise all element values to `missing`. 
```

# Design Document: `of` Type System for JuliaBUGS

## Overview

The `of` type system provides a declarative way to specify parameter **types** for probabilistic programming. It serves as a lightweight type annotation system that:
- Returns actual Julia types (not instances) that can be used in type annotations
- Encodes specifications (dimensions, bounds) in type parameters
- Provides utilities for parameter manipulation

## Core Concepts

### 1. Type-Based Design

The `of` function returns types with specifications encoded in type parameters:
- `of(Array, dims...)` → `OfArray{Float64, N, (dim1, dim2, ...)}` - Arrays with specified dimensions
- `of(Array, T, dims...)` → `OfArray{T, N, (dim1, dim2, ...)}` - Typed arrays
- `of(Real)` → `OfReal{Nothing, Nothing}` - Unbounded real numbers
- `of(Real, lower, upper)` → `OfReal{Val{lower}, Val{upper}}` - Bounded real numbers
- `of(Int)` → `OfInt{Nothing, Nothing}` - Unbounded integers

> - `of(Int, lower, upper)` → `OfInt{Val{lower}, Val{upper}}` - Bounded integers

Reviewer 2: Does this need the `Val`s? Bits types can be type parameters as they are.

- `@of(field1=..., field2=...)` → `OfNamedTuple{(:field1, :field2), Tuple{Type1, Type2}}` - Named tuples (use @of macro only)
- `of(...; constant=true)` → `OfConstantWrapper{T}` - Marks a type as constant/hyperparameter (only supported for `Int` and `Real` types)

### 2. Type Parameter Encoding

The system encodes extra useful information into type parameters:
- **Dimensions**: Stored as tuple type parameters (e.g., `(3, 4)` for a 3×4 matrix)
- **Bounds**: Encoded using `Val{x}` for numeric bounds or `Nothing` for unbounded
- **Symbolic references**: Encoded using `SymbolicRef{:symbol}` for referencing other fields

> - **Arithmetic expressions**: Encoded using `SymbolicExpr{expr}` for expressions like `n+1`, `2*n`, etc. Division operations must result in integers for array dimensions.

Reviewer 2: After reading the examples I see why something like `SymbolicExpr` is needed. It makes me somewhat uncomfortable though, it feels like we are mixing source code (`Expr`s) with values in a way that at least at first seems confusing to me. I wonder if the `Expr`s could somehow be replaced with functions. This isn't so much a comment asking for change, but more a comment flagging something that I don't understand but that seems unusual and possibly troublesome to me.

Reviewer 3: The problem with anon functions is that you can call the parameter `n` but it has no knowledge that that is supposed to be the `n` that came before it. These functions are the same: 

```julia
i -> i + 1
n -> n + 1
```

(cf. alpha conversion in lambda calculus), because the value of `n` isn't the `n` from outside but rather the parameter, so how can we tell whether it's supposed to add one to `i` or `n`? You need to carry a reference to a symbol that is resolved **outside of** the scope of the function. This can only be accomplished with some sort of `Expr` machinery.

Reviewer 2: I'm not sure I understand. The macro knows the names of each of the fields, so it could write the anonymous function as `n -> n + 1` if the first field is called `n`. After macro processing is done, I agree that the distinction between that and `i -> i + 1` is lost, but these functions would always be called with the preceding fields as arguments, in order as positional arguments. So the call would be something like
```julia
constructed_value_for_field_4 = anonymous_functions_for_field_4(val_of_field1, val_of_field2, val_of_field3)
```
Wouldn't that do it?

Reviewer 3: Firstly, sorry, I didn't see your second comment before I posted mine. When I posted it Hong's comment was the last one in the thread (I'm going to blame mobile data on the Tube for that).

> Wouldn't that do it?

Sure, it would, in that the _output_ of the macro would be an anon function. I had in mind that you all wanted the _input_ of the macro to be an anon function, e.g.

```julia
@of(
    halved=(n -> of(Array, n/2, n))
)
```

This wouldn't work for the reasons noted in my comment.

Sure, I think having the macro output anon functions instead of Exprs may be simpler. As you said, it does rely on specific ordering of variables which seems to go against the design suggested elsewhere.

Overall, the point I'm trying to make (and that seems to be lost) is that variable names like this inevitably introduce complexity that has to be dealt with at the Expr level.

There is a fundamental difference between `of(Real)` and `of(Array, n/2, n)`. The first one is an actual Julia function that will actually return something if used outside of the macro. Thus, the macro does not need to _interpret_ `of(Real)`, it can simply pass it through. The second one is inherently nonsensical and it is not possible to write (non-macro) Julia code to give it a meaning. It can only be understood within the context of the `@of` macro.

Regardless of whether the `@of` macro outputs a partially parsed Expr where references to `n` are resolved later, or whether it converts it into an anonymous function where references to `n` are implicitly resolved by `n` being the first parameter, it's still the same amount of complexity.

- `flatten(T::Type{<:OfType}, values)` - Convert structured values to flat vector
- `unflatten(T::Type{<:OfType}, vec)` - Reconstruct structured values from flat vector

### 4. The @of Macro

The `@of` macro provides cleaner syntax by automatically converting field references to symbols:

```julia
# Instead of writing:
T = of((
    n = of(Int; constant=true),
    data = of(Array, :n, 2)
))

# You can write:
T = @of(
    n = of(Int; constant=true),
    data = of(Array, n, 2)  # 'n' is automatically converted to :n
)
```

### 5. Symbolic Dimensions and Bounds

For cases where dimensions need to be specified at runtime:

```julia
# Define type with symbolic dimensions

> MatrixType = of((

Reviewer: I suggest we remove support for using `of((;...))` and only keep `@of()` for composable types for simplicity and to reduce the maintenance burden.

For cases where dimensions need to be specified at runtime:

```julia
# Define type with symbolic dimensions using @of macro
MatrixType = @of(
    rows=of(Int; constant=true),
    cols=of(Int; constant=true),
    data=of(Array, rows, cols),  # Direct references converted to symbols
)
# Note: Expressions like 'rows + 1' are not yet supported

# Create concrete type by specifying constants
ConcreteType = MatrixType(;rows=3, cols=4)
# Can also validate data while concretizing
ConcreteType = MatrixType(;rows=3, cols=4, data=rand(3, 4))  # Validates that data matches 3x4

> data_nt = (data=rand(3, 4),)
> flat = flatten(ConcreteType, data_nt)

Reviewer: Note `ConcreteType` already has randomly initialised `data`, so `data_nt` is unnecessary. 
```suggestion
flat = flatten(ConcreteType)
```

Author: this doesn't feel quite right -- `ConcreteType` in your example should be a type, i.e. 
```julia
ConcreteType == @of(
data = of(Array, 3, 4)
)
```
`flatten` should take a source of values that conforms to the `of` type and returns a vector.

Author: at the moment, as of the last commit

```julia
ConcreteType = MatrixType(;rows=3, cols=4, data=rand(3, 4))  # Validates that data matches 3x4
```
will only check that the size of `data` is consistent with
```julia
MatrixType = @of(
    rows=of(Int; constant=true),
    cols=of(Int; constant=true),
    data=of(Array, rows, cols),  # Direct references converted to symbols
)
```
with
 `rows=3, cols=4`

Reviewer: > will only check that the size of data is consistent 

That would be confusing. I suggest

```julia
x = MatrixType(;rows=3)  # this should error with a message constant `cols` is required but not provided. 
x = MatrixType(;rows=3, cols=4)  # default to MatrixType(;rows=3, cols=4, data = zero(3, 4))
x = MatrixType(;rows=3, cols=4, data = data) # 
```

This would always return an instance when calling `T(;kwargs...)` instead of another type. 
In addition, it requires the user to pass all constants when calling `T(;kwargs...)`, and default to zero if non-constant variables aren't provided explicitly.

Author: this is updated now

    data=of(Array, rows, cols),
)

# Create instance by providing all constants (default to zero for data)
instance = MatrixType(;rows=3, cols=4)  
# instance = (data = zeros(3, 4),)

# Create instance with specific data
instance = MatrixType(;rows=3, cols=4, data=rand(3, 4))  
# instance = (data = <provided 3x4 matrix>,)

# Use flatten/unflatten with unconcretized types (providing constants as kwargs)
flat = flatten(MatrixType, instance; rows=3, cols=4)

> reconstructed = unflatten(MatrixType, flat; rows=3, cols=4)

Reviewer: We can use `of(MatrixType::OfType; rows=3, cols=4)` for creating semi-concrete types, where `rows=3, cols=4` serves a similar role to `lower` and `upper`. 

```suggestion
reconstructed = unflatten(of(MatrixType; rows=3, cols=4), flat; rows=3, cols=4)
```

Note that the first argument of this syntax has to be a subtype of `OfRype`.

Author: yeah, this is already implemented, I can update the design doc

# rand and zero with keyword arguments

> rand(MatrixType; rows=3, cols=4)  # generates random instance
> zero(MatrixType; rows=10, cols=5) # generates zero instance

Reviewer: ```suggestion
rand(of(MatrixType; rows=3, cols=4))  # generates random instance
zero(of(MatrixType; rows=3, cols=4))  # generates zero instance
```

# Missing constants will error
MatrixType(; rows=3) # Error: Constant `cols` is required but not provided

> rand(MatrixType; rows=3) # Error: Missing values for symbolic dimensions: cols

Reviewer: ```suggestion
rand(of(MatrixType; rows=3)) # Error: Missing values for symbolic dimensions: cols
```

    halved=of(Array, n/2, n),
)

# Create instance - all non-constant fields default to zero
instance = ExpandedMatrixType(; n=10)
# This creates an instance with:
# - original: 10×10 zero matrix
# - padded: 11×11 zero matrix
# - doubled: 20×10 zero matrix  
# - halved: 5×10 zero matrix  (n/2 must result in an integer, error if not)

# Generate random instance

> rand(ExpandedMatrixType; n=10)

Reviewer: ```suggestion
rand(of(ExpandedMatrixType; n=10))
```

Reviewer: @sunxd3 For clarity, let's make sure to remove these methods

```julia
zero(ExpandedMatrixType; n=10)
rand(ExpandedMatrixType; n=10)
flat = latten(MatrixType, instance; rows=3, cols=4)
unflatten(MatrixType, flat; rows=3, cols=4)
```

in favour of

```julia
zero(of(ExpandedMatrixType; n=10))
rand(of(ExpandedMatrixType; n=10))
flat = latten(of(MatrixType, instance; rows=3, cols=4))
unflatten(of(MatrixType, flat; rows=3, cols=4))
```

for the principle that there should be only one way of doing things.

Author: done now

    end
end
```

> ### Dynamic Model with Variable Dimensions

Reviewer: ```suggestion
### Model with Variable Dimensions
```

Here's an example of a model where array dimensions depend on other parameters:

```julia

> # Model where the number of mixture components is dynamic

Reviewer: ```suggestion
# Mixture model where the number of mixture components is itself a parameter
```

Tparams = @of(
    n_components=of(Int, 1, 3; constant=true),  # Can be 1, 2, or 3
    weights=of(Array, n_components),            # Size depends on n_components
    means=of(Array, n_components),              # Size depends on n_components

>     y=of(Real)

Reviewer: Let's have `N` data points `y = of(Array, N, 1)`

)

@model function dynamic_mixture((; n_components, weights, means, y)::Tparams)
    # Prior on component probabilities (Dirichlet)
    if n_components == 1
        weights[1] = 1.0  # Single component has weight 1
        means[1] ~ Normal(0.0, 1.0)
        y ~ Normal(means[1], 0.5)
    elseif n_components == 2
        # Two components
        weights ~ Dirichlet([1.0, 1.0])
        means[1] ~ Normal(-1.0, 1.0)
        means[2] ~ Normal(1.0, 1.0)
        # Mixture likelihood
        z ~ Categorical(weights)
        y ~ Normal(means[z], 0.5)
    else  # n_components == 3
        # Three components
        weights ~ Dirichlet([1.0, 1.0, 1.0])
        means[1] ~ Normal(-2.0, 1.0)
        means[2] ~ Normal(0.0, 1.0)
        means[3] ~ Normal(2.0, 1.0)
        # Mixture likelihood
        z ~ Categorical(weights)
        y ~ Normal(means[z], 0.5)
    end
end

# Another example: Variable-order autoregressive model
ARParams = @of(
    order=of(Int, 1, 5; constant=true),  # AR order between 1 and 5
    coeffs=of(Array, order),             # AR coefficients
    sigma=of(Real, 0, nothing),          # Error variance
    y=of(Array, 100),                    # Time series data
)

> @model function dynamic_ar((; order, coeffs, sigma, y)::ARParams, n_obs)

Reviewer: ```suggestion
@model function variable-order_ar((; order, coeffs, sigma, y)::ARParams, n_obs)
```

    # Priors
    for i in 1:order
        coeffs[i] ~ Normal(0.0, 0.5)  # Shrinkage prior on AR coefficients
    end
    sigma ~ InverseGamma(2.0, 1.0)
    
    # AR likelihood
    for t in (order+1):n_obs
        # Compute AR prediction
        pred = 0.0
        for i in 1:order

>             pred += coeffs[i] * y[t-i]

Reviewer: This aligns better with the notation `coeffs' * y` 
```suggestion
            pred += coeffs[order+1-i] * y[t-i]
```

        end
        y[t] ~ Normal(pred, sqrt(sigma))
    end
end

# Usage example
# Concrete type with specific dimensions
ConcreteARType = of(ARParams; order=3)
# This creates a type where coeffs has size 3

data = (
    coeffs=[0.5, -0.3, 0.1],
    sigma=0.25,
    y=randn(100)
)

> params = ConcreteARType(data)

Reviewer: ```suggestion
params = ConcreteARType(;data=data)
```
